### PR TITLE
Nova mm

### PR DESCRIPTION
**Pull request by technionRamy**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

Introducing Nova : a Lock Free Off-Heap Memory Manager.

Nova can be used to replace the SeqExpandMemoryManager or and the SyncRecycleMemoryManager.

### CODE REVIEW

File: core/src/main/java/com/yahoo/oak/BasicChunk.java [hunk 1]
```diff
@@ -241,6 +246,11 @@ boolean overwriteExistingValueForMove(ThreadContext ctx, V newVal) {
     static class Statistics {
         private final AtomicInteger addedCount = new AtomicInteger(0);
         private int initialCount = 0;
+        private static final int VALS_4_AVERAGE_SIZE = 2048;
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/BasicChunk.java:249]: I think it was taken from one of the previous versions of the code. We decided to delete this usage. Please remove unless needed for your code correctness

File: core/src/main/java/com/yahoo/oak/InternalOakMap.java [hunk 2]
```diff
@@ -329,51 +329,55 @@ V put(K key, V value, OakTransformer<V> transformer) {
...
-            OrderedChunk<K, V> c = findChunk(key); // find orderedChunk matching key
-            c.lookUp(ctx, key);
-            // If there is a matching value reference for the given key, and it is not marked as deleted,
-            // then this put changes the slice pointed by this value reference.
-            if (ctx.isValueValid()) {
-                // there is a value and it is not deleted
-                Result res = valueOperator.exchange(c, ctx, value, transformer, valueSerializer);
-                if (res.operationResult == ValueUtils.ValueResult.TRUE) {
-                    return (V) res.value;
+            try {
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/InternalOakMap.java:332]: At least all those tries should be commented, why they are here...

**Review by sanastas — COMMENTED** [review]
Not full review, but something to be discussed tomorrow

File: core/src/main/java/com/yahoo/oak/BasicChunk.java [hunk 3]
```diff
@@ -22,13 +22,15 @@
...
+    protected KeyMemoryManager kMM;
     // in split/compact process, represents parent of split (can be null!)
     private final AtomicReference<BasicChunk<K, V>> creator;
     // chunk can be in the following states: normal, frozen or infant(has a creator)
-    private final AtomicReference<State> state;
+    protected final AtomicReference<State> state;
     private final AtomicReference<Rebalancer<K, V>> rebalancer;
     private final AtomicInteger pendingOps;
     private final int maxItems;
+    protected final boolean releaseKeys;
```
[the remark this answers is not in the document]
Comment (technionRamy) [reply core/src/main/java/com/yahoo/oak/BasicChunk.java:33]: this is to check what MM we are using (to know if we can release keys or not)

File: core/src/main/java/com/yahoo/oak/EntryOrderedSet.java [hunk 4]
```diff
@@ -310,4 +312,21 @@ boolean isEntrySetValidAfterRebalance() {
 
         return true;
     }
+    
+    void releaseAllDeletedKeys() {
+        KeyBuffer key = new KeyBuffer(keysMemoryManager.getEmptySlice());
+        for (int i = 0; i < OrderedChunk.ORDERED_CHUNK_MAX_ITEMS_DEFAULT ; i++) {
+            if (valuesMemoryManager.isReferenceDeleted(getValueReference(i))) {
+                if (keysMemoryManager.isReferenceValidAndNotDeleted(getKeyReference(i))) {
+                    //TODO may add exception to isDelted and stop deleting if key is already deleted
```
[the remark this answers is not in the document]
Comment (technionRamy) [reply core/src/main/java/com/yahoo/oak/EntryOrderedSet.java:321]: Old comment, I removed it!

**Pull request closed without merging**